### DISCUSSION TITLE
Offset Pagination

### DISCUSSION THREAD

**Discussion by khalibloo** [saleor/saleor#12138] [Ideas]
### What I'm trying to achieve
I want to have pagination links in my product list UI. From a UX and SEO perspective, having pagination links is far more preferable to having a "load more" button. The user can reach products at the end of the list with fewer actions/clicks. The same applies to search engine crawlers. A product that requires a dozen clicks to reach will not be easily found by crawlers nor ranked high.

### Describe a proposed solution
Add support for offset pagination in the GraphQL API. There are of course, a lot of performance considerations regarding how the products will be fetched from the database.

**Comment by patrys** [comment]
I am unaware of any performant way to offer generic, real-time page-based pagination combined with custom filters and sorting. Saleor would need to pre-calculate and cache result sets for all possible combinations of filters. If you need page-based navigation, implementing it on top of the API in your backend-for-frontend is the best way.

The easiest way would be to map page numbers to GraphQL filter queries, including `after`, or to map page numbers to object IDs at build time.

The most flexible solution would be to index the products in Algolia, Elasticsearch, Typesense, or Meilisearch and use that for catalog browsing.

**Reply by khalibloo** [reply]
I was afraid that might be the case. I think I'll explore the search service option, since I haven't set up a BFF for my project. Thanks.
